### PR TITLE
Cleaner enlist flow

### PR DESCRIPTION
**Pull request by igor47**
Big PR. A bunch of changes:
* login flow has been factored out into a `LoginOrRegister` object. this object handles the case where you try to register, but your account already exists -- in this case, you'll just be logged in if possible. 
* invite generation factored into an `Invite` object; it correctly handles multiple email addresses with commas or spaces separating them (not sure how this worked before). we have a cannonical place to keep the default invite subject and message body
* we can now see the results of our attempts to invite right in the modal
* there is now a generic invite link (not targetted at any person) and you can select it from the crew page (to invite via offline mechanisms)
* the invite accept flow has been refactored into an `AcceptInvitation` object. the controller correctly handles cases where the user is logged in, the invite was for someone else, or you try to accept the invite with a pre-existing user
* i killed the `rsvp` endpoint, since the `enlist` endpoint now handles it cleanly
* the views code is now a bit smaller, since common functionality was factored out into it's own files/objects.

should fix a lot of @igorkofman 's complaints about logins and the invite/accept flow

### CODE REVIEW

**Comment by igor47** [comment]
invite link box: 
Screenshot from 2019-09-19 12-39-36

File: spaceship/accept_invitation.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+
+from flask import render_template
+
+from spaceship import email
+from spaceship.models import Invitation, Team, TeamUser, User
+
+class AcceptInvitation:
+
+  @classmethod
+  def perform(cls, invitation: Invitation, user: User):
```
Comment (jeredw) [inline spaceship/accept_invitation.py]: Woo, types.  Maybe add -> Team?

File: spaceship/accept_invitation.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    team: Team = invitation.team
+    if user in team.members:
+      return team
```
Comment (staceysv) [inline spaceship/accept_invitation.py]: ah, I guess if they're already on the team it doesn't make a difference

File: spaceship/accept_invitation.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
+
+from flask import render_template
+
+from spaceship import email
+from spaceship.models import Invitation, Team, TeamUser, User
+
+class AcceptInvitation:
```
Comment (igorkofman) [inline spaceship/accept_invitation.py]: the classes don't seem super useful

File: spaceship/invite.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+  @classmethod
+  def default_message(cls, inviter: User, team: Team):
+    return render_template('invite_crew.html', inviter=inviter, team=team)
+
+  @classmethod
+  def perform(cls, inviter: User, team_id: int, subject: str, message: str, emails: List[str]) -> Optional[str]:
+    team = Team.query.get(team_id)
+    if not team or team not in inviter.teams:
+      return f'You are not authorized to invite crew onto team {team_id}'
```
Comment (staceysv) [inline spaceship/invite.py]: if the user is seeing this error message, we should show team name instead of team_id

File: spaceship/invite.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    message = message if message else cls.default_message(inviter, team)
+
+    # quilljs needs <p><br></p> to show line breaks even though everything is in a paragraph
+    # this creates extra line breaks in gmail so just strip it out
+    message = message.replace('<p><br></p>', '')
+
+    if not emails:
+      return f'List a few folks to invite!'
+
+    # an email is anything containing an `@` character, separated by whitespace and/or commas
```
Comment (jeredw) [inline spaceship/invite.py]: this would have been more helpful above EMAIL_REGEX

File: spaceship/invite.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    for invited_email in cls.EMAIL_REGEX.findall(emails):
+      iv = Invitation(
+        inviter=inviter,
+        team=team,
+        invited_email=invited_email,
+        message=message,
+        status='sent')
+      iv.save()
+
+      # each recipient sees a unique invite link
```
Comment (jeredw) [inline spaceship/invite.py]: is this still true?  seems to contradict your pr message

File: spaceship/become_captain.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
+
+from spaceship import names, achievements
+from spaceship.models import Team, TeamUser
+
+class BecomeCaptain:
```
Comment (igorkofman) [inline spaceship/become_captain.py]: create team?

**Review by staceysv — APPROVED** [review]
made some small comments; looks good overall

File: spaceship/models/team.py [hunk 8]
```diff
@@ -20,6 +20,22 @@ class Team(db.Model):
...
+        )
+      )
+    except StopIteration:
+      generic = Invitation(
+        inviter=inviter,
+        team=self,
+      )
+      generic.save()
+
+    return generic
```
Comment (jeredw) [inline spaceship/models/team.py]: nit: i would have found this more readable if written as a for loop that returns inside else does the rest outside

File: spaceship/login_or_register.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+      raise cls.Error('email is required')
+
+    # if the user already exists, log them in
+    u = User.query.filter(User.email == email).first()
+    if u:
+      return cls.login(email, password, oauth_user_info)
+
+    # if we got here, we have to create the user
+    u = User(email=email)
+    u.name = oauth_user_info.get('name', name if name else '')
```
Comment (igorkofman) [inline spaceship/login_or_register.py]: could do email.split('@')[0] for name

File: spaceship/login_or_register.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    u.email_confirmed = email == oauth_user_info.get('email')
+    if password:
+      u.set_password(password)
+    u.save()
+
+    # the user either signed up via 'become a captain' or accepted an invite to join a team
```
Comment (igorkofman) [inline spaceship/login_or_register.py]: is this comment correct?

File: spaceship/templates/base_with_quill.html [hunk 11]
```diff
@@ -108,7 +123,7 @@ <h5 class="modal-title" v-html="modal_title"></h5>
 
       this.savedStates['invite'] = {
         modal_title: "Invite Crew",
-        subject: "Join my Spaceship Earth crew!",
+        subject: "{{ invitation_subject }}",
```
Comment (jeredw) [inline spaceship/templates/base_with_quill.html:126]: escaped for js?

File: spaceship/models/team.py [hunk 12]
```diff
@@ -20,6 +20,22 @@ class Team(db.Model):
   created_at = db.Column(PendulumDateTimeField(), default=lambda: pendulum.now('UTC'))
   deleted_at = db.Column(PendulumDateTimeField(), nullable=True)
 
+  def generic_invitation_from(self, inviter):
```
Comment (igorkofman) [inline spaceship/models/team.py]: this seems really fancy.  maybe a comment to explain?

File: spaceship/models/team.py [hunk 13]
```diff
@@ -20,6 +20,22 @@ class Team(db.Model):
...
+    try:
+      generic = next(
+        i for i in self.invitations if (
+          i.inviter == inviter and i.invited_email is None and not i.already_accepted
```
Comment (igorkofman) [inline spaceship/models/team.py]: why not already_accepted?

**Review by svetlichnaya — APPROVED** [review]

**Review by jeredw — APPROVED** [review]
i didn't look really closely at the views.py changes but on a cursory read this seems much neater.  however i am somewhat fried so probably worth banging on it some more...

File: spaceship/invite.py (hunk 6 above)
Comment (igor47) [reply spaceship/invite.py]: yeah, we still send folks unique invite links which will auto-populate the email address field in the registration. but we ALSO make a non-customized link available.

File: spaceship/login_or_register.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+
+  # log in using a password?
+  elif password and user.check_password(password):
+    return user
+
+  # out of options, cannot log in
+  else:
+    raise LoginFailed()
+
+def perform(email=None, name=None, password=None, oauth_user_info={}, as_captain=False):
```
Comment (igorkofman) [inline spaceship/login_or_register.py]: this isn't a great method name

**Review by igorkofman — APPROVED** [review]
so great!

File: spaceship/templates/base_with_quill.html (hunk 11 above)
Comment (igor47) [reply spaceship/templates/base_with_quill.html:126]: mmm the default is html-safe, so html entities will show up ugly. is html-safe a superset of js-safe?

File: spaceship/models/team.py (hunk 13 above)
Comment (igor47) [reply spaceship/models/team.py]: not sure. we should probably kill invite statuses altogether

File: spaceship/accept_invitation.py (hunk 3 above)
Comment (igor47) [reply spaceship/accept_invitation.py]: killed them all with :fire:

File: spaceship/become_captain.py (hunk 7 above)
Comment (igor47) [reply spaceship/become_captain.py]: done

File: spaceship/login_or_register.py (hunk 9 above)
Comment (igor47) [reply spaceship/login_or_register.py]: done!

File: spaceship/login_or_register.py (hunk 10 above)
Comment (igor47) [reply spaceship/login_or_register.py]: fixed the comment

File: spaceship/login_or_register.py (hunk 14 above)
Comment (igor47) [reply spaceship/login_or_register.py]: renamed

File: spaceship/templates/base_with_quill.html (hunk 11 above)
Comment (jeredw) [reply spaceship/templates/base_with_quill.html:126]: i was more thinking about ".  random people on the internet say to do (no quotes) {{ invitation_subject|tojson }} in jinja2

File: spaceship/invite.py (hunk 6 above)
Comment (jeredw) [reply spaceship/invite.py]: https://i.kym-cdn.com/photos/images/original/000/538/731/0fc.gif

**Pull request merged**